scripts/bert/bert_qa_evaluate.py

```diff
@@ -0,0 +1,409 @@
+
+        total_scores = []
+        best_non_null_entry = None
+        for entry in nbest:
+            total_scores.append(entry.start_logit + entry.end_logit)
+            if not best_non_null_entry:
+                if entry.text:
+                    best_non_null_entry = entry
+
+        probs = nd.softmax(nd.array(total_scores)).asnumpy()
+
+        nbest_json = []
+        for (i, entry) in enumerate(nbest):
+            output = OrderedDict()
+            output['text'] = entry.text
+            output['probability'] = float(probs[i])
+            output['start_logit'] = entry.start_logit
+            output['end_logit'] = entry.end_logit
+            nbest_json.append(output)
+
+        if not version_2:
+            all_predictions[example_qas_id] = nbest_json[0]['text']
+        else:
+            # predict '' iff the null score - the score of best non-null > threshold
+            score_diff = score_null - best_non_null_entry.start_logit - \
+                best_non_null_entry.end_logit
+
+            scores_diff_json[example_qas_id] = score_diff
+            if score_diff > null_score_diff_threshold:
+                all_predictions[example_qas_id] = ''
+            else:
+                all_predictions[example_qas_id] = best_non_null_entry.text
+
+        all_nbest_json[example_qas_id] = nbest_json
+    return all_predictions, all_nbest_json, scores_diff_json
+
+
+def normalize_answer(s):
+    """Lower text and remove punctuation, articles and extra whitespace."""
+
+    def remove_articles(text):
+        return re.sub(r'\b(a|an|the)\b', ' ', text)
+
+    def white_space_fix(text):
+        return ' '.join(text.split())
+
+    def remove_punc(text):
+        exclude = set(string.punctuation)
+        return ''.join(ch for ch in text if ch not in exclude)
+
+    def lower(text):
+        return text.lower()
+
+    return white_space_fix(remove_articles(remove_punc(lower(s))))
+
+
+def f1_score(prediction, ground_truth):
+    """Calculate the F1 scores.
+    """
+    prediction_tokens = normalize_answer(prediction).split()
+    ground_truth_tokens = normalize_answer(ground_truth).split()
+    common = Counter(prediction_tokens) & Counter(ground_truth_tokens)
+    num_same = sum(common.values())
+    if num_same == 0:
+        return 0
+    precision = 1.0 * num_same / len(prediction_tokens)
+    recall = 1.0 * num_same / len(ground_truth_tokens)
+    f1 = (2 * precision * recall) / (precision + recall)
+    return f1
+
+
+def exact_match_score(prediction, ground_truth):
+    """Calculate the EM scores.
+    """
+    return (normalize_answer(prediction) == normalize_answer(ground_truth))
+
+
+def metric_max_over_ground_truths(metric_fn, prediction, ground_truths):
+    scores_for_ground_truths = []
+    for ground_truth in ground_truths:
+        score = metric_fn(prediction, ground_truth)
+        scores_for_ground_truths.append(score)
+    return max(scores_for_ground_truths)
+
+
+def get_F1_EM(dataset, predict_data):
```

Comment:
I forget to mention that, as the evaluation code is adopted from official evaluation script, it would be better if we can cite their code: https://worksheets.codalab.org/rest/bundles/0xbcd57bee090b421c982906709c8c27e1/contents/blob/

Reply:
When I was looking for the `evaluate-v1.1` script, I didn't find the official script.
But there are scripts from https://github.com/allenai/bi-att-flow/blob/master/squad/evaluate-v1.1.py in the original repo.
For the sake of convenience, I merged the script code into `bert_qa_evaluate.py`.
Is it necessary to provide the original repo script address or official script address.